pulp_ansible/app/tasks/copy.py

```diff
@@ -62,3 +73,80 @@ def process_entry(entry):
         with dest_repo.new_version(base_version=base_version) as new_version:
             new_version.add_content(content_to_copy)
+
+
+def copy_collection(cv_pk_list, src_repo_pk, dest_repo_list):
+    """
+    Copy a collection and all of it's contents into a list of destination repositories
+    """
+    src_repo = AnsibleRepository.objects.get(pk=src_repo_pk)
+    collection_versions = CollectionVersion.objects.filter(pk__in=cv_pk_list)
+
+    content_types = src_repo.content.values_list("pulp_type", flat=True).distinct()
+    source_pks = src_repo.content.values_list("pk", flat=True)
+
+    content = cv_pk_list
+
+    # collection signatures
+    if "ansible.collection_signature" in content_types:
+        signatures_pks = CollectionVersionSignature.objects.filter(
+            signed_collection__in=cv_pk_list, pk__in=source_pks
+        ).values_list("pk", flat=True)
+        if signatures_pks:
+            content.extend(signatures_pks)
+
+    # collection version mark
+    if "ansible.collection_mark" in content_types:
+        marks_pks = CollectionVersionMark.objects.filter(
+            marked_collection__in=cv_pk_list, pk__in=source_pks
+        ).values_list("pk", flat=True)
+        if marks_pks:
+            content.extend(marks_pks)
+
+    # namespace metadata
+    namespaces = {x.namespace for x in collection_versions}
+    namespaces_pks = AnsibleNamespaceMetadata.objects.filter(
+        pk__in=source_pks,
+        name__in=list(namespaces),
+    ).values_list("pk", flat=True)
+    if namespaces_pks:
+        content.extend(namespaces_pks)
+
+    # collection deprecation
+    for cv in collection_versions:
+        if "ansible.collection_deprecation" in content_types:
+            deprecations_pks = AnsibleCollectionDeprecated.objects.filter(
+                pk__in=source_pks,
+                namespace=cv.namespace,
+                name=cv.name,
+            ).values_list("pk", flat=True)
+            if deprecations_pks:
+                content.extend(deprecations_pks)
+
+    for pk in dest_repo_list:
+        add_and_remove(repository_pk=pk, add_content_units=content, remove_content_units=[])
+
+
+def move_collection(cv_pk_list, src_repo_pk, dest_repo_list):
+    """
+    Copy a collection and all of it's contents into a list of destination repositories
+    """
+    copy_collection(cv_pk_list, src_repo_pk, dest_repo_list)
+
+    # don't need to clean anything other than the collection version because everything
+    # else will get handled by AnsibleRepository.finalize_repo_version()
+    add_and_remove(repository_pk=src_repo_pk, add_content_units=[], remove_content_units=cv_pk_list)
+
+
+def copy_or_move_and_sign(copy_or_move, signing_service_pk=None, **move_task_kwargs):
+    if signing_service_pk:
+        sign(
+            signing_service_href=signing_service_pk,
+            repository_href=move_task_kwargs["src_repo_pk"],
+            content_hrefs=move_task_kwargs["cv_pk_list"],
+        )
```

Comment:
This will sign the things in the original Repository, and not purely "on the fly", right?
This strikes me as a rather non-intuitive side-effect. I might be convinced otherwise.

Reply:
I didn't want the overhead of calling sign on each dest repository. If there's a way to sign the collections without specifying a repo, and then add the signatures as a separate operation, I can do that.

Reply:
And we def don't want different signatures on all dest repos. Would it be wise to see if there was a signature already? Or is the `sign` call taking care of that?
Yes, i'd prefer keeping the source repo untouched. You'd probably need to rewrite a `sign` method that instead of adding to a repo returns the resulting pk(s).

Reply:
`sign` will grab the signatures already present in Pulp instead of resigning, so calling sign on each dest repo is fine. The first repo will perform the actual signing and everyone after will add the generated signatures. The only annoying thing will be all the extra progress report messages that will be present in the task report.

Reply:
@gerrod3 @mdellweg I updated it to call sign on each dest repo. LMK if you want me to leave it like this, or change it.

Reply:
This will create a whole lot of new repository versions, no?